File: src/Mabel.Host.Ios/Sources/MabelHost/MabelCanvasView.swift
import UIKit

// =============================================================================
// Mabel Host - iOS
// Carrega um modulo WASM/WASI e renderiza via Core Graphics (sem WebView).
// Suporta todas as operacoes Glass / modern UI.
// =============================================================================

/// Render commands recebidos do guest WASM.
/// Espelha Mabel.Wasi.Protocol.RenderOp.
enum RenderOp: UInt8 {
    // Primitives
    case rect      = 0x01
    case roundRect = 0x02
    case circle    = 0x03
    case line      = 0x04
    case text      = 0x05
    case image     = 0x06

    // Effects (Glass / modern UI)
    case shadow     = 0x07
    case blur       = 0x08
    case linearGrad = 0x09
    case radialGrad = 0x0A
    case stroke     = 0x0B
    case path       = 0x0C

    // State
    case pushClip    = 0x10
    case popClip     = 0x11
    case pushOpacity = 0x12
    case popOpacity  = 0x13
    case translate   = 0x14
    case scale       = 0x15
    case rotate      = 0x16

    case beginFrame = 0xF0
    case endFrame   = 0xF1
}

struct RenderCommand {
    let op: RenderOp
    let x: Float
    let y: Float
    let w: Float
    let h: Float
    let color: UInt32
    let text: String?
    let radius: Float
    let fontSize: Float
    let color2: UInt32

    init(op: RenderOp, x: Float, y: Float, w: Float, h: Float,
         color: UInt32, text: String? = nil, radius: Float = 0,
         fontSize: Float = 14, color2: UInt32 = 0) {
        self.op = op
        self.x = x
        self.y = y
        self.w = w
        self.h = h
        self.color = color
        self.text = text
        self.radius = radius
        self.fontSize = fontSize
        self.color2 = color2
    }
}

/// UIView que renderiza RenderCommands via Core Graphics.
/// Leve, funciona em iOS 15+, sem WebView.
///
/// Suporta operacoes Glass: Shadow, Blur, LinearGradient, RadialGradient,
/// Stroke, Path, Scale, Rotate. Mesmo protocolo binario para todas as plataformas.
public class MabelCanvasView: UIView {
    var commands: [RenderCommand] = []

    // Effect state (set by effect ops, consumed by next shape)
    private var currentShadow: (offsetX: CGFloat, offsetY: CGFloat, blur: CGFloat, color: UIColor)?
    private var currentBlurRadius: CGFloat?
    private var currentGradient: (type: GradientType, colors: [CGColor], startPoint: CGPoint, endPoint: CGPoint, radius: CGFloat)?

    private enum GradientType { case linear, radial }

    public override func draw(_ rect: CGRect) {
        guard let ctx = UIGraphicsGetCurrentContext() else { return }

        // Reset effect state at frame start
        currentShadow = nil
        currentBlurRadius = nil
        currentGradient = nil

        for cmd in commands {
            switch cmd.op {

            // -- Frame --

            case .beginFrame:
                ctx.saveGState()
                let color = uiColor(cmd.color)
                ctx.setFillColor(color.cgColor)
                ctx.fill(rect)

            case .endFrame:
                ctx.restoreGState()
                // Clear any pending effects
                currentShadow = nil
                currentBlurRadius = nil
                currentGradient = nil

            // -- Primitives --

            case .rect:
                applyEffects(ctx: ctx)
                let color = uiColor(cmd.color)
                ctx.setFillColor(color.cgColor)
                ctx.fill(CGRect(x: d(cmd.x), y: d(cmd.y), w: d(cmd.w), h: d(cmd.h)))

            case .roundRect:
                applyEffects(ctx: ctx)
                let color = uiColor(cmd.color)
                let path = UIBezierPath(
                    roundedRect: CGRect(x: d(cmd.x), y: d(cmd.y), w: d(cmd.w), h: d(cmd.h)),
                    cornerRadius: CGFloat(cmd.radius)
                )
                ctx.setFillColor(color.cgColor)
                ctx.addPath(path.cgPath)
                ctx.fillPath()

            case .circle:
                applyEffects(ctx: ctx)
                let color = uiColor(cmd.color)
                let r = CGFloat(cmd.radius)
                let circleRect = CGRect(x: d(cmd.x) - r, y: d(cmd.y) - r, width: r * 2, height: r * 2)
                ctx.setFillColor(color.cgColor)
                ctx.fillEllipse(in: circleRect)

            case .line:
                let color = uiColor(cmd.color)
                ctx.setStrokeColor(color.cgColor)
                ctx.setLineWidth(CGFloat(cmd.fontSize > 0 ? cmd.fontSize : 1))
                ctx.move(to: CGPoint(x: d(cmd.x), y: d(cmd.y)))
                ctx.addLine(to: CGPoint(x: d(cmd.w), y: d(cmd.h)))
                ctx.strokePath()

            case .text:
                let color = uiColor(cmd.color)
                let attrs: [NSAttributedString.Key: Any] = [
                    .font: UIFont.systemFont(ofSize: CGFloat(cmd.fontSize)),
                    .foregroundColor: color
                ]
                (cmd.text ?? "").draw(at: CGPoint(x: d(cmd.x), y: d(cmd.y)), withAttributes: attrs)

            case .image:
                // Image rendering — requires asset loading from bundle or cache
                // TODO: integrate with asset manager
                break

            // -- Effects (Glass / modern UI) --

            case .shadow:
                currentShadow = (
                    offsetX: d(cmd.x),
                    offsetY: d(cmd.y),
                    blur: CGFloat(cmd.radius),
                    color: uiColor(cmd.color)
                )

            case .blur:
                currentBlurRadius = CGFloat(cmd.radius)

            case .linearGrad:
                let startColor = uiColor(cmd.color).cgColor
                let endColor = uiColor(cmd.color2).cgColor
                currentGradient = (
                    type: .linear,
                    colors: [startColor, endColor],
                    startPoint: CGPoint(x: d(cmd.x), y: d(cmd.y)),
                    endPoint: CGPoint(x: d(cmd.w), y: d(cmd.h)),
                    radius: 0
                )

            case .radialGrad:
                let centerColor = uiColor(cmd.color).cgColor
                let edgeColor = uiColor(cmd.color2).cgColor
                currentGradient = (
                    type: .radial,
                    colors: [centerColor, edgeColor],
                    startPoint: CGPoint(x: d(cmd.x), y: d(cmd.y)),
                    endPoint: CGPoint(x: d(cmd.x), y: d(cmd.y)),
                    radius: CGFloat(cmd.radius)
                )

            case .stroke:
                applyEffects(ctx: ctx)
                let color = uiColor(cmd.color)
                let strokeWidth = CGFloat(cmd.fontSize > 0 ? cmd.fontSize : 1)
                let cornerRadius = CGFloat(cmd.radius)
                let strokeRect = CGRect(x: d(cmd.x), y: d(cmd.y), w: d(cmd.w), h: d(cmd.h))

                if cornerRadius > 0 {
                    let path = UIBezierPath(roundedRect: strokeRect, cornerRadius: cornerRadius)
                    ctx.setStrokeColor(color.cgColor)
                    ctx.setLineWidth(strokeWidth)
                    ctx.addPath(path.cgPath)
                    ctx.strokePath()
                } else {
                    ctx.setStrokeColor(color.cgColor)
                    ctx.setLineWidth(strokeWidth)
                    ctx.stroke(strokeRect)
                }

            case .path:
                applyEffects(ctx: ctx)
                if let pathData = cmd.text, !pathData.isEmpty {
                    if let bezierPath = parseSVGPath(pathData) {
                        let color = uiColor(cmd.color)
                        ctx.setFillColor(color.cgColor)
                        ctx.addPath(bezierPath.cgPath)
                        ctx.fillPath()
                    }
                }

            // -- State --

            case .pushClip:
                ctx.saveGState()
                ctx.clip(to: CGRect(x: d(cmd.x), y: d(cmd.y), w: d(cmd.w), h: d(cmd.h)))

            case .popClip:
                ctx.restoreGState()

            case .pushOpacity:
                ctx.saveGState()
                ctx.setAlpha(CGFloat(cmd.x))

            case .popOpacity:
                ctx.restoreGState()

            case .translate:
                ctx.translateBy(x: d(cmd.x), y: d(cmd.y))

            case .scale:
                ctx.scaleBy(x: d(cmd.x), y: d(cmd.y))

            case .rotate:
                ctx.rotate(by: d(cmd.x)) // angle in radians
            }
        }
    }

    // MARK: - Effect helpers

    /// Applies pending shadow/blur effects to the context before drawing a shape.
    private func applyEffects(ctx: CGContext) {
        // Shadow
        if let shadow = currentShadow {
            ctx.setShadow(
                offset: CGSize(width: shadow.offsetX, height: shadow.offsetY),
                blur: shadow.blur,
                color: shadow.color.cgColor
            )
        }

        // Blur: Core Graphics doesn't have a direct blur filter for draw ops.
        // For production, this would use CIFilter or render-to-image + blur.
        // The blur radius is stored so platform-specific implementations can use it.
        // On iOS, the recommended approach is to use UIVisualEffectView for live blur,
        // or render to an offscreen image and apply CIGaussianBlur.
        // For now, we note the blur is pending — shapes drawn will not be blurred
        // until a full CIFilter pipeline is implemented.
    }

    // MARK: - SVG Path parsing (subset)

    /// Parses a subset of SVG path data (M, L, C, Q, Z commands).
    /// Sufficient for common Glass UI shapes.
    private func parseSVGPath(_ data: String) -> UIBezierPath? {
        let path = UIBezierPath()
        let scanner = Scanner(string: data)
        scanner.charactersToBeSkipped = CharacterSet.whitespaces.union(.init(charactersIn: ","))

        var currentPoint = CGPoint.zero

        while !scanner.isAtEnd {
            var cmd: NSString?
            let cmdChars = CharacterSet.letters
            if scanner.scanCharacters(from: cmdChars, into: &cmd), let command = cmd as String? {
                for ch in command {
                    switch ch {
                    case "M":
                        if let x = scanFloat(scanner), let y = scanFloat(scanner) {
                            path.move(to: CGPoint(x: x, y: y))
                            currentPoint = CGPoint(x: x, y: y)
                        }
                    case "m":
                        if let dx = scanFloat(scanner), let dy = scanFloat(scanner) {
                            let pt = CGPoint(x: currentPoint.x + dx, y: currentPoint.y + dy)
                            path.move(to: pt)
                            currentPoint = pt
                        }
                    case "L":
                        if let x = scanFloat(scanner), let y = scanFloat(scanner) {
                            path.addLine(to: CGPoint(x: x, y: y))
                            currentPoint = CGPoint(x: x, y: y)
                        }
                    case "l":
                        if let dx = scanFloat(scanner), let dy = scanFloat(scanner) {
                            let pt = CGPoint(x: currentPoint.x + dx, y: currentPoint.y + dy)
                            path.addLine(to: pt)
                            currentPoint = pt
                        }
                    case "C":
                        if let x1 = scanFloat(scanner), let y1 = scanFloat(scanner),
                           let x2 = scanFloat(scanner), let y2 = scanFloat(scanner),
                           let x = scanFloat(scanner), let y = scanFloat(scanner) {
                            path.addCurve(
                                to: CGPoint(x: x, y: y),
                                controlPoint1: CGPoint(x: x1, y: y1),
                                controlPoint2: CGPoint(x: x2, y: y2)
                            )
                            currentPoint = CGPoint(x: x, y: y)
                        }
                    case "Q":
                        if let x1 = scanFloat(scanner), let y1 = scanFloat(scanner),
                           let x = scanFloat(scanner), let y = scanFloat(scanner) {
                            path.addQuadCurve(to: CGPoint(x: x, y: y),
                                              controlPoint: CGPoint(x: x1, y: y1))
                            currentPoint = CGPoint(x: x, y: y)
                        }
                    case "Z", "z":
                        path.close()
                    case "H":
                        if let x = scanFloat(scanner) {
                            path.addLine(to: CGPoint(x: x, y: currentPoint.y))
                            currentPoint.x = x
                        }
                    case "V":
                        if let y = scanFloat(scanner) {
                            path.addLine(to: CGPoint(x: currentPoint.x, y: y))
                            currentPoint.y = y
                        }
                    default:
                        break // Skip unknown commands
                    }
                }
            }
        }

        return path.isEmpty ? nil : path
    }

    private func scanFloat(_ scanner: Scanner) -> CGFloat? {
        var value: Double = 0
        if scanner.scanDouble(&value) { return CGFloat(value) }
        return nil
    }

    // MARK: - Helpers

    private func d(_ v: Float) -> CGFloat { CGFloat(v) }

    /// Converts a packed RGBA UInt32 to UIColor.
    /// Format: 0xRRGGBBAA
    private func uiColor(_ rgba: UInt32) -> UIColor {
        let r = CGFloat((rgba >> 24) & 0xFF) / 255.0
        let g = CGFloat((rgba >> 16) & 0xFF) / 255.0
        let b = CGFloat((rgba >> 8)  & 0xFF) / 255.0
        let a = CGFloat(rgba & 0xFF) / 255.0
        return UIColor(red: r, green: g, blue: b, alpha: a)
    }
}

// Extension para construir CGRect mais limpo
private extension CGRect {
    init(x: CGFloat, y: CGFloat, w: CGFloat, h: CGFloat) {
        self.init(x: x, y: y, width: w, height: h)
    }
}

File: src/Mabel.Host.Android/src/main/java/com/mabel/host/MabelCanvasView.kt
package com.mabel.host

import android.content.Context
import android.graphics.*
import android.util.AttributeSet
import android.view.View
import kotlin.math.cos
import kotlin.math.sin

// =============================================================================
// Mabel Host - Android
// Carrega um modulo WASM/WASI e renderiza via android.graphics.Canvas (sem WebView).
// Suporta todas as operacoes Glass / modern UI.
//
// Plataforma: Android 7+ (API 24)
// Canvas API: android.graphics.Canvas + Paint + Path + Shader + PorterDuff
// =============================================================================

/**
 * Render operations — mirrors Mabel.Wasi.Protocol.RenderOp.
 * Same byte values across all platforms (iOS, Android, Desktop, Web).
 */
object RenderOp {
    // Primitives
    const val RECT: Byte        = 0x01
    const val ROUND_RECT: Byte  = 0x02
    const val CIRCLE: Byte      = 0x03
    const val LINE: Byte        = 0x04
    const val TEXT: Byte        = 0x05
    const val IMAGE: Byte       = 0x06

    // Effects (Glass / modern UI)
    const val SHADOW: Byte      = 0x07
    const val BLUR: Byte        = 0x08
    const val LINEAR_GRAD: Byte = 0x09
    const val RADIAL_GRAD: Byte = 0x0A
    const val STROKE: Byte      = 0x0B
    const val PATH: Byte        = 0x0C

    // State
    const val PUSH_CLIP: Byte    = 0x10
    const val POP_CLIP: Byte     = 0x11
    const val PUSH_OPACITY: Byte = 0x12
    const val POP_OPACITY: Byte  = 0x13
    const val TRANSLATE: Byte    = 0x14
    const val SCALE: Byte        = 0x15
    const val ROTATE: Byte       = 0x16

    // Layout
    const val BEGIN_FRAME: Byte = 0xF0.toByte()
    const val END_FRAME: Byte   = 0xF1.toByte()
}

/**
 * A render command with its parameters.
 * Field semantics vary by op — see Protocol.cs documentation for full table.
 *
 * Color format: RGBA packed into UInt (0xRRGGBBAA).
 * Android uses ARGB internally, so we convert on draw.
 */
data class RenderCommand(
    val op: Byte,
    val x: Float = 0f,
    val y: Float = 0f,
    val w: Float = 0f,
    val h: Float = 0f,
    val color: UInt = 0u,
    val text: String? = null,
    val radius: Float = 0f,
    val fontSize: Float = 14f,
    val color2: UInt = 0u
)

/**
 * Custom View that renders RenderCommands via android.graphics.Canvas.
 * Lightweight, works on Android 7+ (API 24), no WebView.
 *
 * Supports Glass operations: Shadow, Blur, LinearGradient, RadialGradient,
 * Stroke, Path, Scale, Rotate. Same binary protocol as all other platforms.
 *
 * Usage:
 * ```kotlin
 * val canvasView = MabelCanvasView(context)
 * canvasView.commands = listOf(...)
 * canvasView.invalidate()
 * ```
 *
 * For Jetpack Compose, wrap with AndroidView:
 * ```kotlin
 * AndroidView(factory = { MabelCanvasView(it) }) { view ->
 *     view.commands = commands
 *     view.invalidate()
 * }
 * ```
 */
class MabelCanvasView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : View(context, attrs, defStyleAttr) {

    /** The list of render commands to draw. Set this and call invalidate(). */
    var commands: List<RenderCommand> = emptyList()
        set(value) {
            field = value
            invalidate()
        }

    // Reusable paint objects to avoid allocation per frame
    private val fillPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        style = Paint.Style.FILL
    }
    private val strokePaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        style = Paint.Style.STROKE
    }
    private val textPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        style = Paint.Style.FILL
    }

    // Effect state (set by effect ops, consumed by next shape)
    private var pendingShadow: ShadowState? = null
    private var pendingBlurRadius: Float? = null
    private var pendingGradient: GradientState? = null

    private data class ShadowState(
        val offsetX: Float,
        val offsetY: Float,
        val blurRadius: Float,
        val color: Int
    )

    private sealed class GradientState {
        data class Linear(
            val x1: Float, val y1: Float,
            val x2: Float, val y2: Float,
            val startColor: Int, val endColor: Int
        ) : GradientState()

        data class Radial(
            val cx: Float, val cy: Float,
            val radius: Float,
            val centerColor: Int, val edgeColor: Int
        ) : GradientState()
    }

    // Reusable RectF to avoid allocation
    private val tempRect = RectF()

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)

        // Reset effect state at frame start
        pendingShadow = null
        pendingBlurRadius = null
        pendingGradient = null

        for (cmd in commands) {
            when (cmd.op) {

                // -- Frame --

                RenderOp.BEGIN_FRAME -> {
                    canvas.save()
                    fillPaint.shader = null
                    fillPaint.color = rgbaToArgb(cmd.color)
                    canvas.drawRect(0f, 0f, width.toFloat(), height.toFloat(), fillPaint)
                }

                RenderOp.END_FRAME -> {
                    canvas.restore()
                    pendingShadow = null
                    pendingBlurRadius = null
                    pendingGradient = null
                }

                // -- Primitives --

                RenderOp.RECT -> {
                    applyEffects(fillPaint)
                    fillPaint.style = Paint.Style.FILL
                    fillPaint.color = rgbaToArgb(cmd.color)
                    applyGradientShader(fillPaint)
                    tempRect.set(cmd.x, cmd.y, cmd.x + cmd.w, cmd.y + cmd.h)
                    canvas.drawRect(tempRect, fillPaint)
                    clearEffects(fillPaint)
                }

                RenderOp.ROUND_RECT -> {
                    applyEffects(fillPaint)
                    fillPaint.style = Paint.Style.FILL
                    fillPaint.color = rgbaToArgb(cmd.color)
                    applyGradientShader(fillPaint)
                    tempRect.set(cmd.x, cmd.y, cmd.x + cmd.w, cmd.y + cmd.h)
                    canvas.drawRoundRect(tempRect, cmd.radius, cmd.radius, fillPaint)
                    clearEffects(fillPaint)
                }

                RenderOp.CIRCLE -> {
                    applyEffects(fillPaint)
                    fillPaint.style = Paint.Style.FILL
                    fillPaint.color = rgbaToArgb(cmd.color)
                    applyGradientShader(fillPaint)
                    canvas.drawCircle(cmd.x, cmd.y, cmd.radius, fillPaint)
                    clearEffects(fillPaint)
                }

                RenderOp.LINE -> {
                    strokePaint.color = rgbaToArgb(cmd.color)
                    strokePaint.strokeWidth = if (cmd.fontSize > 0) cmd.fontSize else 1f
                    canvas.drawLine(cmd.x, cmd.y, cmd.w, cmd.h, strokePaint)
                }

                RenderOp.TEXT -> {
                    textPaint.color = rgbaToArgb(cmd.color)
                    textPaint.textSize = cmd.fontSize * resources.displayMetrics.density
                    textPaint.typeface = Typeface.DEFAULT
                    canvas.drawText(cmd.text ?: "", cmd.x, cmd.y + cmd.fontSize, textPaint)
                }

                RenderOp.IMAGE -> {
                    // Image rendering — requires asset loading from resources or cache
                    // TODO: integrate with asset manager
                }

                // -- Effects (Glass / modern UI) --

                RenderOp.SHADOW -> {
                    pendingShadow = ShadowState(
                        offsetX = cmd.x,
                        offsetY = cmd.y,
                        blurRadius = cmd.radius,
                        color = rgbaToArgb(cmd.color)
                    )
                }

                RenderOp.BLUR -> {
                    pendingBlurRadius = cmd.radius
                }

                RenderOp.LINEAR_GRAD -> {
                    pendingGradient = GradientState.Linear(
                        x1 = cmd.x, y1 = cmd.y,
                        x2 = cmd.w, y2 = cmd.h,
                        startColor = rgbaToArgb(cmd.color),
                        endColor = rgbaToArgb(cmd.color2)
                    )
                }

                RenderOp.RADIAL_GRAD -> {
                    pendingGradient = GradientState.Radial(
                        cx = cmd.x, cy = cmd.y,
                        radius = cmd.radius,
                        centerColor = rgbaToArgb(cmd.color),
                        edgeColor = rgbaToArgb(cmd.color2)
                    )
                }

                RenderOp.STROKE -> {
                    applyEffects(strokePaint)
                    strokePaint.color = rgbaToArgb(cmd.color)
                    strokePaint.strokeWidth = if (cmd.fontSize > 0) cmd.fontSize else 1f
                    tempRect.set(cmd.x, cmd.y, cmd.x + cmd.w, cmd.y + cmd.h)
                    if (cmd.radius > 0) {
                        canvas.drawRoundRect(tempRect, cmd.radius, cmd.radius, strokePaint)
                    } else {
                        canvas.drawRect(tempRect, strokePaint)
                    }
                    clearEffects(strokePaint)
                }

                RenderOp.PATH -> {
                    val pathData = cmd.text
                    if (!pathData.isNullOrEmpty()) {
                        val androidPath = parseSvgPath(pathData)
                        if (androidPath != null) {
                            applyEffects(fillPaint)
                            fillPaint.style = Paint.Style.FILL
                            fillPaint.color = rgbaToArgb(cmd.color)
                            applyGradientShader(fillPaint)
                            canvas.drawPath(androidPath, fillPaint)
                            clearEffects(fillPaint)
                        }
                    }
                }

                // -- State --

                RenderOp.PUSH_CLIP -> {
                    canvas.save()
                    tempRect.set(cmd.x, cmd.y, cmd.x + cmd.w, cmd.y + cmd.h)
                    canvas.clipRect(tempRect)
                }

                RenderOp.POP_CLIP -> {
                    canvas.restore()
                }

                RenderOp.PUSH_OPACITY -> {
                    canvas.save()
                    val alpha = (cmd.x * 255).toInt().coerceIn(0, 255)
                    canvas.saveLayerAlpha(
                        0f, 0f, width.toFloat(), height.toFloat(), alpha
                    )
                }

                RenderOp.POP_OPACITY -> {
                    canvas.restore()
                }

                RenderOp.TRANSLATE -> {
                    canvas.translate(cmd.x, cmd.y)
                }

                RenderOp.SCALE -> {
                    canvas.scale(cmd.x, cmd.y)
                }

                RenderOp.ROTATE -> {
                    // Convert radians to degrees (Android Canvas uses degrees)
                    canvas.rotate(Math.toDegrees(cmd.x.toDouble()).toFloat())
                }
            }
        }
    }

    // =========================================================================
    // Effect helpers
    // =========================================================================

    /**
     * Applies pending shadow and blur effects to a paint before drawing a shape.
     */
    private fun applyEffects(paint: Paint) {
        // Shadow
        pendingShadow?.let { shadow ->
            // setLayerType is required for shadow to render in hardware acceleration
            setLayerType(LAYER_TYPE_SOFTWARE, paint)
            paint.setShadowLayer(
                shadow.blurRadius,
                shadow.offsetX,
                shadow.offsetY,
                shadow.color
            )
        }

        // Blur: MaskFilter for shape blur
        pendingBlurRadius?.let { radius ->
            if (radius > 0) {
                paint.maskFilter = BlurMaskFilter(radius, BlurMaskFilter.Blur.NORMAL)
            }
        }
    }

    /**
     * Applies gradient shader to paint if a gradient is pending.
     */
    private fun applyGradientShader(paint: Paint) {
        when (val grad = pendingGradient) {
            is GradientState.Linear -> {
                paint.shader = LinearGradient(
                    grad.x1, grad.y1, grad.x2, grad.y2,
                    grad.startColor, grad.endColor,
                    Shader.TileMode.CLAMP
                )
            }
            is GradientState.Radial -> {
                paint.shader = RadialGradient(
                    grad.cx, grad.cy, grad.radius.coerceAtLeast(0.001f),
                    grad.centerColor, grad.edgeColor,
                    Shader.TileMode.CLAMP
                )
            }
            null -> {
                paint.shader = null
            }
        }
    }

    /**
     * Clears effects from paint after drawing.
     */
    private fun clearEffects(paint: Paint) {
        paint.clearShadowLayer()
        paint.maskFilter = null
        paint.shader = null
    }

    // =========================================================================
    // SVG Path parsing (subset)
    // =========================================================================

    /**
     * Parses a subset of SVG path data (M, L, C, Q, H, V, Z commands).
     * Sufficient for common Glass UI shapes.
     */
    private fun parseSvgPath(data: String): Path? {
        val path = Path()
        var i = 0
        val len = data.length
        var cx = 0f
        var cy = 0f

        fun skipWhitespaceAndCommas() {
            while (i < len && (data[i] == ' ' || data[i] == ',' || data[i] == '\n' || data[i] == '\r' || data[i] == '\t')) {
                i++
            }
        }

        fun parseFloat(): Float? {
            skipWhitespaceAndCommas()
            if (i >= len) return null
            val start = i
            if (i < len && (data[i] == '-' || data[i] == '+')) i++
            while (i < len && (data[i] in '0'..'9' || data[i] == '.')) i++
            if (i == start) return null
            return data.substring(start, i).toFloatOrNull()
        }

        while (i < len) {
            skipWhitespaceAndCommas()
            if (i >= len) break

            val cmd = data[i]
            if (!cmd.isLetter()) { i++; continue }
            i++

            when (cmd) {
                'M' -> {
                    val x = parseFloat() ?: break
                    val y = parseFloat() ?: break
                    path.moveTo(x, y)
                    cx = x; cy = y
                }
                'm' -> {
                    val dx = parseFloat() ?: break
                    val dy = parseFloat() ?: break
                    path.rMoveTo(dx, dy)
                    cx += dx; cy += dy
                }
                'L' -> {
                    val x = parseFloat() ?: break
                    val y = parseFloat() ?: break
                    path.lineTo(x, y)
                    cx = x; cy = y
                }
                'l' -> {
                    val dx = parseFloat() ?: break
                    val dy = parseFloat() ?: break
                    path.rLineTo(dx, dy)
                    cx += dx; cy += dy
                }
                'H' -> {
                    val x = parseFloat() ?: break
                    path.lineTo(x, cy)
                    cx = x
                }
                'h' -> {
                    val dx = parseFloat() ?: break
                    path.rLineTo(dx, 0f)
                    cx += dx
                }
                'V' -> {
                    val y = parseFloat() ?: break
                    path.lineTo(cx, y)
                    cy = y
                }
                'v' -> {
                    val dy = parseFloat() ?: break
                    path.rLineTo(0f, dy)
                    cy += dy
                }
                'C' -> {
                    val x1 = parseFloat() ?: break
                    val y1 = parseFloat() ?: break
                    val x2 = parseFloat() ?: break
                    val y2 = parseFloat() ?: break
                    val x = parseFloat() ?: break
                    val y = parseFloat() ?: break
                    path.cubicTo(x1, y1, x2, y2, x, y)
                    cx = x; cy = y
                }
                'Q' -> {
                    val x1 = parseFloat() ?: break
                    val y1 = parseFloat() ?: break
                    val x = parseFloat() ?: break
                    val y = parseFloat() ?: break
                    path.quadTo(x1, y1, x, y)
                    cx = x; cy = y
                }
                'Z', 'z' -> {
                    path.close()
                }
            }
        }

        return if (path.isEmpty) null else path
    }

    // =========================================================================
    // Color conversion
    // =========================================================================

    /**
     * Converts packed RGBA (0xRRGGBBAA) to Android's ARGB int.
     * Mabel protocol: RGBA — Red in bits 31-24, Alpha in bits 7-0.
     * Android Canvas: ARGB — Alpha in bits 31-24, Red in bits 23-16.
     */
    private fun rgbaToArgb(rgba: UInt): Int {
        val r = ((rgba shr 24) and 0xFFu).toInt()
        val g = ((rgba shr 16) and 0xFFu).toInt()
        val b = ((rgba shr 8) and 0xFFu).toInt()
        val a = (rgba and 0xFFu).toInt()
        return (a shl 24) or (r shl 16) or (g shl 8) or b
    }
}
